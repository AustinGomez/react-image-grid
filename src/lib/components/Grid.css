.grid-img.lightbox-enabled {
    cursor: pointer;
}
