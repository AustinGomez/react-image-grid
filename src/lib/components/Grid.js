import React, { useState, useEffect, useCallback } from 'react'
import PropTypes from 'prop-types'
import useWindowSize from '../hooks/UseWindowSize'
import Lightbox from './Lightbox'
import './Grid.css'

const Grid = React.forwardRef(({ images, rowHeight, padding = 0 }, ref) => {
    let windowSize = useWindowSize()
    const [minAspectRatio, setMinAspectRatio] = useState()
    const [rows, setRows] = useState([])
    const [showLightbox, setShowLightbox] = useState(false)
    const [selectedIndex, setSelectedIndex] = useState()



    // We'll use this value to calculate how many pictures we need in a row.
    // The "min" aspect ratio is the aspect ratio that will allow the row to
    // span the correct length while being between minRowHeight and maxRowHeight
    const calcMinAspectRatio = useCallback(() => {
        return (ref.current.parentNode.clientWidth) / rowHeight
    }, [rowHeight, ref])

    const handleImageClick = (e) => {
        setSelectedIndex(e.target.getAttribute('data-index'))
        setShowLightbox(true)
    }

    // Creates <img/> tags from image and row metadata.
    // 
    const imageDataToImgTag = useCallback((image, totalIndex, rowIndex, row, width) => {
        const calculatedRowHeight = ref.current.parentNode.clientWidth / width

        // Calculate the dimensions and padding of each image. This needs
        // to be inline since we need some values from the JS.
        let imageStyle = {
            height: calculatedRowHeight + "px",

            // Take back out the padding from the ratio.
            width: calculatedRowHeight * image[1] - padding + "px",

            // Don't add padding to the last item in a row
            paddingRight: rowIndex === row.length - 1 ? 0 : padding + "px"
        }
        return <img
            className="grid-img"
            style={imageStyle}
            data-index={totalIndex}
            onClick={handleImageClick}
            src={image[0].download_url}
            alt={image[0].alt}
            key={"img_" + image[0].id + "_" + image[1]} />
    }, [ref, padding])


    // Build the rows of the grid. Each row must have an aspect ratio of at least minAspectRatio.
    // Then, each image in the row is scaled up to fill the desired width of the row, while maintaining
    // the aspect ratio of each photo in the row.
    const makeRows = useCallback(() => {
        let allRows = []
        let row = []
        let width = 0
        for (let i = 0; i < images.length; i++) {
            let image = images[i]

            // Add the padding into the ratio.
            let ratio = image.width / image.height

            // If we're less than the min aspectRatio then keep adding more items to the row.
            if (width <= minAspectRatio && i !== images.length - 1) {
                // Add the ratio contributed by the padding.
                ratio += padding / image.height
                row.push([image, ratio])
                width += ratio
            }
            else {
                if (i === images.length - 1) {
                    row.push([image, ratio])
                    width += ratio
                }
                let imageElements = []
                for (let j = 0; j < row.length; j++) {
                    imageElements.push(imageDataToImgTag(row[j], i - row.length + j, j, row, width))
                }
                // Add the finished row to the list of all rows.
                allRows.push(imageElements)

                // Start a new row with the current image as the first image
                row = [[image, ratio]]
                width = ratio
            }
        }

        // This style also needs to be inline since we use the padding prop.
        // (Would be nice to use something like styled-components for this instead)
        const divStyle = {
            display: "flex",
            paddingBottom: padding + "px"
        }
        setRows(allRows.map((row, index) => <div className="grid-row" style={divStyle} key={"row_" + index}>{row}</div>))
    }, [images, padding, minAspectRatio, imageDataToImgTag])

    // Recalculate min aspect ratio when the window size changes.
    useEffect(() => {
        setMinAspectRatio(calcMinAspectRatio())
    }, [calcMinAspectRatio, windowSize])

    // This isn't optimal. Ideally we don't want to call makeRows every time the aspect ratio changes.
    // There should be a resizeRows method that just goes through each row and recalculates and resizes
    // each row/image.
    useEffect(() => {
        if (minAspectRatio) {
            makeRows()
        }
    }, [minAspectRatio, makeRows])

    return (
        <div>
            <div className="grid-container" ref={ref}>
                {rows}
            </div>
            {showLightbox ?
                <Lightbox
                    selectedImage={images[selectedIndex]}
                    nextImage={images[(selectedIndex + 1) % images.length]}
                    prevImage={images[(selectedIndex + images.length - 1) % images.length]}
                    onClose={() => setShowLightbox(false)}
                    onPrev={() => setSelectedIndex(selectedIndex => (selectedIndex + images.length - 1) % images.length)}
                    onNext={() => setSelectedIndex(selectedIndex => (selectedIndex + 1) % images.length)} />
                : null
            }
        </div>
    )
})

Grid.propTypes = {
    images: PropTypes.arrayOf(
        PropTypes.shape({
            download_url: PropTypes.string.isRequired,
            height: PropTypes.number.isRequired,
            width: PropTypes.number.isRequired
        })).isRequired,
    rowHeight: PropTypes.number.isRequired,
    padding: PropTypes.number
}

export default Grid