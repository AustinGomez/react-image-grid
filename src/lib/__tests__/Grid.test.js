import Grid from "../components/Grid";
import React from "react";
import { shallow, mount, render } from "enzyme";

describe("Grid component", () => {
  let images;
  let wrapper;
  beforeEach(() => {
    images = [
      {
        src: "fake",
        width: 100,
        height: 100
      }
    ];

    wrapper = shallow(<Grid images={images} rowHeight={100} width={500} />);
  });
  it("should render no images", () => {
    images = [];
    wrapper = shallow(<Grid images={images} rowHeight={100} width={500} />);
    expect(wrapper.find(".grid-container")).toHaveLength(1);
    expect(wrapper.find(".grid-row")).toHaveLength(0);
    expect(wrapper.find(".grid-img")).toHaveLength(0);
    expect(wrapper).toMatchSnapshot();
  });

  it("should render one image", () => {
    expect(wrapper.find(".grid-container")).toHaveLength(1);
    expect(wrapper.find(".grid-row")).toHaveLength(1);
    expect(wrapper.find(".grid-img")).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  it("should render many images", () => {
    images = [
      { src: "https://via.placeholder.com/371x463", width: 3710, height: 4630 },
      { src: "https://via.placeholder.com/278x587", width: 2780, height: 5870 },
      { src: "https://via.placeholder.com/364x415", width: 3640, height: 4150 },
      { src: "https://via.placeholder.com/490x463", width: 4900, height: 4630 },
      { src: "https://via.placeholder.com/480x490", width: 4800, height: 4900 },
      { src: "https://via.placeholder.com/214x551", width: 2140, height: 5510 },
      { src: "https://via.placeholder.com/669x506", width: 6690, height: 5060 },
      { src: "https://via.placeholder.com/509x583", width: 5090, height: 5830 },
      { src: "https://via.placeholder.com/320x362", width: 3200, height: 3620 },
      { src: "https://via.placeholder.com/304x449", width: 3040, height: 4490 }
    ];
    wrapper = shallow(<Grid images={images} rowHeight={100} width={500} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should respond to different widths", () => {
    images = [
      { src: "https://via.placeholder.com/371x463", width: 3710, height: 4630 },
      { src: "https://via.placeholder.com/278x587", width: 2780, height: 5870 },
      { src: "https://via.placeholder.com/364x415", width: 3640, height: 4150 },
      { src: "https://via.placeholder.com/490x463", width: 4900, height: 4630 },
      { src: "https://via.placeholder.com/480x490", width: 4800, height: 4900 },
      { src: "https://via.placeholder.com/214x551", width: 2140, height: 5510 },
      { src: "https://via.placeholder.com/669x506", width: 6690, height: 5060 },
      { src: "https://via.placeholder.com/509x583", width: 5090, height: 5830 },
      { src: "https://via.placeholder.com/320x362", width: 3200, height: 3620 },
      { src: "https://via.placeholder.com/304x449", width: 3040, height: 4490 }
    ];
    wrapper = shallow(<Grid images={images} rowHeight={100} width={900} />);
    expect(wrapper).toMatchSnapshot();
  });
});
