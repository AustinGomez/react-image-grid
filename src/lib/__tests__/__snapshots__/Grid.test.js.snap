// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid component should match snapshot with no images 1`] = `
<Fragment>
  <div
    className="grid-container"
  />
</Fragment>
`;

exports[`Grid component should match snapshot with one image 1`] = `
<Fragment>
  <div
    className="grid-container"
  >
    <div
      className="grid-row"
      key="row_0"
      style={
        Object {
          "display": "flex",
          "marginBottom": "0px",
        }
      }
    >
      <img
        className="grid-img"
        data-index={-1}
        key="img_undefined_1"
        onClick={[Function]}
        src="fake"
        style={
          Object {
            "height": "500px",
            "marginRight": 0,
            "width": "500px",
          }
        }
      />
    </div>
  </div>
</Fragment>
`;

exports[`Grid component should render many images 1`] = `
<Fragment>
  <div
    className="grid-container"
  >
    <div
      className="grid-row"
      key="row_0"
      style={
        Object {
          "display": "flex",
          "marginBottom": "0px",
        }
      }
    >
      <img
        className="grid-img"
        data-index={0}
        key="img_undefined_0.8012958963282938"
        onClick={[Function]}
        src="https://via.placeholder.com/371x463"
        style={
          Object {
            "height": "84.73965581979495px",
            "marginRight": "0px",
            "width": "67.9015384646737px",
          }
        }
      />
      <img
        className="grid-img"
        data-index={1}
        key="img_undefined_0.4735945485519591"
        onClick={[Function]}
        src="https://via.placeholder.com/278x587"
        style={
          Object {
            "height": "84.73965581979495px",
            "marginRight": "0px",
            "width": "40.13223904242418px",
          }
        }
      />
      <img
        className="grid-img"
        data-index={2}
        key="img_undefined_0.8771084337349397"
        onClick={[Function]}
        src="https://via.placeholder.com/364x415"
        style={
          Object {
            "height": "84.73965581979495px",
            "marginRight": "0px",
            "width": "74.32586679133821px",
          }
        }
      />
      <img
        className="grid-img"
        data-index={3}
        key="img_undefined_1.0583153347732182"
        onClick={[Function]}
        src="https://via.placeholder.com/490x463"
        style={
          Object {
            "height": "84.73965581979495px",
            "marginRight": "0px",
            "width": "89.68127721749357px",
          }
        }
      />
      <img
        className="grid-img"
        data-index={4}
        key="img_undefined_0.9795918367346939"
        onClick={[Function]}
        src="https://via.placeholder.com/480x490"
        style={
          Object {
            "height": "84.73965581979495px",
            "marginRight": "0px",
            "width": "83.01027508877873px",
          }
        }
      />
      <img
        className="grid-img"
        data-index={5}
        key="img_undefined_0.38838475499092556"
        onClick={[Function]}
        src="https://via.placeholder.com/214x551"
        style={
          Object {
            "height": "84.73965581979495px",
            "marginRight": "0px",
            "width": "32.91159046358642px",
          }
        }
      />
      <img
        className="grid-img"
        data-index={6}
        key="img_undefined_1.3221343873517786"
        onClick={[Function]}
        src="https://via.placeholder.com/669x506"
        style={
          Object {
            "height": "84.73965581979495px",
            "marginRight": 0,
            "width": "112.03721293170517px",
          }
        }
      />
    </div>
    <div
      className="grid-row"
      key="row_1"
      style={
        Object {
          "display": "flex",
          "marginBottom": "0px",
        }
      }
    >
      <img
        className="grid-img"
        data-index={6}
        key="img_undefined_0.8730703259005146"
        onClick={[Function]}
        src="https://via.placeholder.com/509x583"
        style={
          Object {
            "height": "205.4140268360489px",
            "marginRight": "0px",
            "width": "179.34089135428624px",
          }
        }
      />
      <img
        className="grid-img"
        data-index={7}
        key="img_undefined_0.8839779005524862"
        onClick={[Function]}
        src="https://via.placeholder.com/320x362"
        style={
          Object {
            "height": "205.4140268360489px",
            "marginRight": "0px",
            "width": "181.58146018656254px",
          }
        }
      />
      <img
        className="grid-img"
        data-index={8}
        key="img_undefined_0.6770601336302895"
        onClick={[Function]}
        src="https://via.placeholder.com/304x449"
        style={
          Object {
            "height": "205.4140268360489px",
            "marginRight": 0,
            "width": "139.07764845915113px",
          }
        }
      />
    </div>
  </div>
</Fragment>
`;

exports[`Grid component should respond to different widths 1`] = `
<Fragment>
  <div
    className="grid-container"
  >
    <div
      className="grid-row"
      key="row_0"
      style={
        Object {
          "display": "flex",
          "marginBottom": "0px",
        }
      }
    >
      <img
        className="grid-img"
        data-index={-1}
        key="img_undefined_0.8012958963282938"
        onClick={[Function]}
        src="https://via.placeholder.com/371x463"
        style={
          Object {
            "height": "107.98444739894734px",
            "marginRight": "0px",
            "width": "86.527494568055px",
          }
        }
      />
      <img
        className="grid-img"
        data-index={0}
        key="img_undefined_0.4735945485519591"
        onClick={[Function]}
        src="https://via.placeholder.com/278x587"
        style={
          Object {
            "height": "107.98444739894734px",
            "marginRight": "0px",
            "width": "51.14084561653724px",
          }
        }
      />
      <img
        className="grid-img"
        data-index={1}
        key="img_undefined_0.8771084337349397"
        onClick={[Function]}
        src="https://via.placeholder.com/364x415"
        style={
          Object {
            "height": "107.98444739894734px",
            "marginRight": "0px",
            "width": "94.71406952582369px",
          }
        }
      />
      <img
        className="grid-img"
        data-index={2}
        key="img_undefined_1.0583153347732182"
        onClick={[Function]}
        src="https://via.placeholder.com/490x463"
        style={
          Object {
            "height": "107.98444739894734px",
            "marginRight": "0px",
            "width": "114.28159659931792px",
          }
        }
      />
      <img
        className="grid-img"
        data-index={3}
        key="img_undefined_0.9795918367346939"
        onClick={[Function]}
        src="https://via.placeholder.com/480x490"
        style={
          Object {
            "height": "107.98444739894734px",
            "marginRight": "0px",
            "width": "105.78068316631575px",
          }
        }
      />
      <img
        className="grid-img"
        data-index={4}
        key="img_undefined_0.38838475499092556"
        onClick={[Function]}
        src="https://via.placeholder.com/214x551"
        style={
          Object {
            "height": "107.98444739894734px",
            "marginRight": "0px",
            "width": "41.93951314587065px",
          }
        }
      />
      <img
        className="grid-img"
        data-index={5}
        key="img_undefined_1.3221343873517786"
        onClick={[Function]}
        src="https://via.placeholder.com/669x506"
        style={
          Object {
            "height": "107.98444739894734px",
            "marginRight": "0px",
            "width": "142.7699512053276px",
          }
        }
      />
      <img
        className="grid-img"
        data-index={6}
        key="img_undefined_0.8730703259005146"
        onClick={[Function]}
        src="https://via.placeholder.com/509x583"
        style={
          Object {
            "height": "107.98444739894734px",
            "marginRight": "0px",
            "width": "94.27801668278593px",
          }
        }
      />
      <img
        className="grid-img"
        data-index={7}
        key="img_undefined_0.8839779005524862"
        onClick={[Function]}
        src="https://via.placeholder.com/320x362"
        style={
          Object {
            "height": "107.98444739894734px",
            "marginRight": "0px",
            "width": "95.45586510404185px",
          }
        }
      />
      <img
        className="grid-img"
        data-index={8}
        key="img_undefined_0.6770601336302895"
        onClick={[Function]}
        src="https://via.placeholder.com/304x449"
        style={
          Object {
            "height": "107.98444739894734px",
            "marginRight": 0,
            "width": "73.11196438592425px",
          }
        }
      />
    </div>
  </div>
</Fragment>
`;
